### PR TITLE
Introduce TypeMetadataWalker and use it in PropertyValidationAccess

### PR DESCRIPTION
**Pull request by asodja**
This is a cherry-pick of most of the work from https://github.com/gradle/gradle/pull/22788.

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test QFL

File: subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalker.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        }
+
+        visitor.visitNested(typeMetadata, qualifiedName, propertyMetadata, node);
+        walkChildren(node, typeMetadata, qualifiedName, visitor, nestedNodesOnPath);
+        nestedNodesOnPath.remove(node);
+    }
+
+    abstract void onNestedNodeCycle(@Nullable String firstOccurrenceQualifiedName, String secondOccurrenceQualifiedName);
+
+    abstract void handleProvider(T node, Consumer<T> handler);
```
Comment (lptr) [inline subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalker.java:94]: ❓ Why aren't we using `protected` here like we do for all the other `abstract` methods?

File: subprojects/model-core/src/main/java/org/gradle/internal/schema/PropertySchemaExtractor.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.schema;
+
+public interface PropertySchemaExtractor {
+}
```
Comment (lptr) [inline subprojects/model-core/src/main/java/org/gradle/internal/schema/PropertySchemaExtractor.java:20]: :x: This type shouldn't be here in this PR.

File: subprojects/persistent-cache/src/testFixtures/groovy/org/gradle/cache/internal/TestCrossBuildInMemoryCacheFactory.groovy [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2018 the original author or authors.
+ * Copyright 2022 the original author or authors.
```
Comment (lptr) [inline subprojects/persistent-cache/src/testFixtures/groovy/org/gradle/cache/internal/TestCrossBuildInMemoryCacheFactory.groovy:2]: 🧐 Shouldn't be updated.

**Comment by asodja** [comment]
@bot-gradle test QFL

File: subprojects/model-core/src/test/groovy/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalkerTest.groovy [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+            null,
+            "nested", "nested.secondNested", "nested.secondNested.thirdNested", "nested.secondNested.thirdNested.input",
+            "nestedList.*", "nestedList.*.secondNested", "nestedList.*.secondNested.thirdNested", "nestedList.*.secondNested.thirdNested.input",
+            "nestedListOfLists.*.*", "nestedListOfLists.*.*.secondNested", "nestedListOfLists.*.*.secondNested.thirdNested", "nestedListOfLists.*.*.secondNested.thirdNested.input",
+            "nestedMap.<key>", "nestedMap.<key>.secondNested", "nestedMap.<key>.secondNested.thirdNested", "nestedMap.<key>.secondNested.thirdNested.input",
+            "nestedProperty", "nestedProperty.secondNested", "nestedProperty.secondNested.thirdNested", "nestedProperty.secondNested.thirdNested.input",
+        ]
+    }
+
+    def "instance walker should handle instances with nested cycles for '#propertyWithCycle' property"() {
```
[the remark this answers is not in the document]
Comment (asodja) [reply subprojects/model-core/src/test/groovy/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalkerTest.groovy]: Should be fixed now

File: subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalker.java (hunk 1 above)
Comment (asodja) [reply subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalker.java]: Overlooked, should be fixed now.

File: subprojects/model-core/src/main/java/org/gradle/internal/schema/PropertySchemaExtractor.java (hunk 2 above)
Comment (asodja) [reply subprojects/model-core/src/main/java/org/gradle/internal/schema/PropertySchemaExtractor.java]: Removed.

File: subprojects/persistent-cache/src/testFixtures/groovy/org/gradle/cache/internal/TestCrossBuildInMemoryCacheFactory.groovy (hunk 3 above)
Comment (asodja) [reply subprojects/persistent-cache/src/testFixtures/groovy/org/gradle/cache/internal/TestCrossBuildInMemoryCacheFactory.groovy]: Fixed.

File: subprojects/model-core/src/test/groovy/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalkerTest.groovy [hunk 5]
```diff
@@ -31,121 +32,109 @@ import java.util.function.Supplier
...
+            "nested::$nestedType",
+            "nestedList.\$1::$nestedType",
+            "nestedList.\$2::$nestedType",
+            "nestedListOfLists.\$1.\$1::$nestedType",
+            "nestedMap.key1::$nestedType",
+            "nestedMap.key2::$nestedType",
+            "nestedNamedList.\$1::$namedType",
+            "nestedProperty::$nestedType"
+        ] as List<String>
+        visitor.leavesToString == [
```
Comment (lptr) [inline subprojects/model-core/src/test/groovy/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalkerTest.groovy:118]: 🤔 These properties would look better without the `ToString` suffix.

**Review by lptr — APPROVED** [review]
LGTM.

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by lptr** [comment]
@bot-gradle test and merge please.

**Pull request merged**